### PR TITLE
feat: Allow from_period and to_period to be timestamps in CSV export functionality

### PR DESCRIPTION
**Pull request by vbaranov**
Resolves https://github.com/blockscout/blockscout/issues/10733

## Motivation

Allow from_period and to_period to be timestamps in CSV export functionality

## Changelog

`from_period` and `to_period` is now possible to pass as timestamps in [ISO 8601:2019](https://en.wikipedia.org/wiki/ISO_8601) format e.g. `2025-02-12T19:43:00Z`. The date format is supported for backward compatibility as well. So, `2025-02-12` input is still legit.

Refactoring of CSV export functionality.

### AI Agent summary

This pull request includes significant refactoring and cleanup of the CSV export functionality, as well as updates to the routing and controller logic. The changes mainly focus on renaming modules, moving CSV export logic to a new controller, and updating routes to reflect these changes.

Refactoring and renaming:

* Renamed `CSVExportController` to `CsvExportController` and updated all references accordingly. [[1]](diffhunk://#diff-bfefcdf17eb1671e1c12403cd464fa7572dc906289d7f14fb6a0b389815a6464L1-R7) [[2]](diffhunk://#diff-bfefcdf17eb1671e1c12403cd464fa7572dc906289d7f14fb6a0b389815a6464L28-R28) [[3]](diffhunk://#diff-bfefcdf17eb1671e1c12403cd464fa7572dc906289d7f14fb6a0b389815a6464L50-R50)
* Renamed CSV export modules and updated their references, such as `AddressInternalTransactionCsvExporter` to `Address.InternalTransactions`. [[1]](diffhunk://#diff-62a862518313d9adc4683094358dde0eb2a5c64be928e01dec1500bf35e82150L1-R8) [[2]](diffhunk://#diff-aec929ea84072ada48b7be221b08f23312a2df929c05814c4b81a22f7e71fd61L1-R8) [[3]](diffhunk://#diff-db66dc15b2721584d650e761e3412c9ded4afebf3b8bcd6171b9ee912cbbb399L1-R1) [[4]](diffhunk://#diff-db66dc15b2721584d650e761e3412c9ded4afebf3b8bcd6171b9ee912cbbb399L16-R16)

Controller and routing updates:

* Removed CSV export-related functions from `AddressTransactionController` and added them to `AddressController`. [[1]](diffhunk://#diff-54264c24cf6c35ae7bdb4a54eaed647c25c39c75ad7e7d41f636762411063a95L172-L236) [[2]](diffhunk://#diff-72707e0a7def9fe96cc51be955265e529bca2292af1f7c5792851ac8bab95565R1087-R1151)
* Updated routes in `ApiRouter` to use the new CSV export endpoints under `/api/v2/**` path and marked the old endpoints for future removal. [[1]](diffhunk://#diff-7b31aa9ca8af99c00b6548ddcef75c78c1824715a3cd6d5e10dbcc6ace229780R220-R226) [[2]](diffhunk://#diff-7b31aa9ca8af99c00b6548ddcef75c78c1824715a3cd6d5e10dbcc6ace229780R236) [[3]](diffhunk://#diff-7b31aa9ca8af99c00b6548ddcef75c78c1824715a3cd6d5e10dbcc6ace229780L490-R500)

Other updates:

* Updated import and alias statements to reflect the new module names and paths. [[1]](diffhunk://#diff-492bf2b89f54211b8d293f53ba30d9899f570dca8d9435eb68bfa2748bdde5fbL8-R11) [[2]](diffhunk://#diff-492bf2b89f54211b8d293f53ba30d9899f570dca8d9435eb68bfa2748bdde5fbL90-R99) [[3]](diffhunk://#diff-d8416289099300803da1b34941165e83189ca4e55cde36c67d35e71fc5abcb14L7-R7) [[4]](diffhunk://#diff-d8416289099300803da1b34941165e83189ca4e55cde36c67d35e71fc5abcb14L22-R22)
* Modified the `CsvExportView` and related templates to use the new CSV export paths. [[1]](diffhunk://#diff-b2457ecf88c523a03a33cf675463afdc379d7cd490ec7c589fc37c8be21f2a29L30-R30) [[2]](diffhunk://#diff-d2237b24f1055d078b982fd7e4bad063707130909e3ed5c9b428e5597b5a277fL337-R337)

## Checklist for your Pull Request (PR)

- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [ ] I checked whether I should update the docs and did so by submitting a PR to [docs repository](https://github.com/blockscout/docs).
- [ ] If I added/changed/removed ENV var, I submitted a PR to [docs repository](https://github.com/blockscout/docs) to update the list of [env vars](https://github.com/blockscout/docs/blob/master/setup/env-variables/README.md) and I updated the version to `master` in the Version column. If I removed variable, I added it to [Deprecated ENV Variables](https://github.com/blockscout/docs/blob/master/setup/env-variables/deprecated-env-variables/README.md) page. After merging docs PR, changes will be reflected in these [pages](https://docs.blockscout.com/setup/env-variables).
- [ ] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [ ] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

## Summary by CodeRabbit

- **New Features**
  - Introduced API v2 endpoints for CSV exports covering transactions, token transfers, internal transactions, logs, and election rewards.
  - Updated UI download links to point to the new CSV export paths.

- **Refactor**
  - Consolidated CSV export functionality by replacing legacy endpoints with a streamlined, consistently named implementation.
  - Improved query logic for block number filtering to include edge cases.

- **Tests**
  - Removed outdated CSV export tests and added new test suites reflecting the updated export behavior.
  - Updated test cases to align with new naming conventions and improved date handling logic.

### CODE REVIEW

**Review by sl1depengwyn — APPROVED** [review]

File: apps/explorer/lib/explorer/chain/csv_export/helper.ex [hunk 1]
```diff
@@ -30,13 +32,80 @@ defmodule Explorer.Chain.CSVExport.Helper do
...
+  end
+
+  defp date_string_to_timestamp_string(date_string) do
+    if date?(date_string) do
+      date_string <> "T00:00:00Z"
+    else
+      date_string
+    end
+  end
+
```
Comment (fedor-ivn) [inline apps/explorer/lib/explorer/chain/csv_export/helper.ex:108]: nitpick: added spec and deduplicate code (didn't test though)

```suggestion
    from_block = convert_date_string_to_block(from_period, :after)
    to_block = convert_date_string_to_block(to_period, :before)

    {from_block, to_block}
  end
  
  @spec convert_date_string_to_block(String.t(), :before | :after) :: integer()
  defp convert_date_string_to_block(date_string, direction) do
    with {:ok, timestamp, _utc_offset} <- date_string_to_timestamp(date_string),
         {:ok, min_block} <- BlockGeneralReader.timestamp_to_block_number(timestamp, direction, true) do
      min_block
    else
      _ -> 0
    end
  end

  @spec date_string_to_timestamp(String.t()) ::
          {:ok, DateTime.t(), Calendar.utc_offset()} | {:error, atom()}
  defp date_string_to_timestamp(date_string) do
    date_string
    |> Date.from_iso8601()
    |> case do
      {:ok, date} ->
        date_string <> "T00:00:00Z"

      _ ->
        date_string
    end
    |> DateTime.from_iso8601()
  end
```

Comment (vbaranov) [reply apps/explorer/lib/explorer/chain/csv_export/helper.ex]: Processed in https://github.com/blockscout/blockscout/pull/11862/commits/a4784e60d5c9f95f02fc8380b7cade6b4c897720.

File: apps/explorer/lib/explorer/chain.ex [hunk 2]
```diff
@@ -419,6 +419,46 @@ defmodule Explorer.Chain do
...
+      # Returns the base query without any filtering
+      ```
+
+    - When both `from_block` and `to_block` are not `nil`:
+      ```elixir
+      where_block_number_in_period(query, 50, 100)
+      # Filters the query to include records with block_number between 50 and 100 (inclusive)
+      ```
+  """
+  @spec where_block_number_in_period(Ecto.Query.t(), non_neg_integer() | nil, non_neg_integer() | nil) :: Ecto.Query.t()
```
[the remark this answers is not in the document]
Comment (vbaranov) [reply apps/explorer/lib/explorer/chain.ex]: It makes sense. Moved to `Explorer.Chain.Block.Reader.General` https://github.com/blockscout/blockscout/pull/11862/commits/115019ffa7b42f7787bcb101f53d49569ed3a0d6.

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex [hunk 3]
```diff
@@ -5,10 +5,10 @@ defmodule BlockScoutWeb.API.V2.AdvancedFilterController do
   import Explorer.PagingOptions, only: [default_paging_options: 0]
 
   alias BlockScoutWeb.CaptchaHelper
-  alias BlockScoutWeb.API.V2.{AdvancedFilterView, CSVExportController}
+  alias BlockScoutWeb.API.V2.{AdvancedFilterView, CsvExportController}
   alias Explorer.{Chain, PagingOptions}
   alias Explorer.Chain.{AdvancedFilter, ContractMethod, Data, Token, Transaction}
-  alias Explorer.Chain.CSVExport.Helper, as: CSVHelper
+  alias Explorer.Chain.CsvExport.Helper, as: CsvHelper
```
[the remark this answers is not in the document]
Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex:11]: I have mixed feelings about your suggestion. I agree with standardization of acronyms. However, I don't like the idea to have namings like this `CSVHelper`, because this introduces an uncertaintanty about where abbreveature is finished and new word started. As for my personal taste, naming like this `CSV.Helper` is fine but this requires a deeper refactoring, and I wouldn't do this in the scope of the current PR. Current PR removes ambigiouty in naming: "Csv" and "CSV" are merged to a single pattern "Csv". So, if you don't mind, I'd suggest you to create a separate PR with migration to acronyms to uppercase, and please refrane from naming like `CSVHelper`, `DBConnection` etc.

[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex:11]: Sure, my main goal was to highlight the importance of naming consistency. So, no additional refactoring is needed in this PR.

I completely agree with your point about names like `CSVHelper`. This is a valid concern, and I think it would be beneficial to hold a broader discussion on a standardized approach to handling acronyms across our modules so that everyone has input.

I can create a follow-up issue on this topic and manage it further. What do you think?

[the remark this answers is not in the document]
Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex:11]: Yes, that sounds great to me. Please go ahead.

**Review by fedor-ivn — APPROVED** [review]
